In [1]:
import pandas as pd
import numpy as np
import statsmodels as st
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [2]:
from scipy import stats

In [2]:
df = pd.read_excel('mult_regr.xlsx')

In [3]:
df.head()

HeatFlux   East  South  North
0     271.8  33.53  40.55  16.66
1     264.0  36.50  36.19  16.46
2     238.8  34.66  37.31  17.66
3     230.7  33.13  32.52  17.50
4     251.6  35.75  33.71  16.40

In [4]:
df.shape

(29, 4)

In [7]:
flux = df[df.columns[0]]

In [9]:
flux.head()

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [10]:
flux[:4]

0    271.8
1    264.0
2    238.8
3    230.7
Name: HeatFlux, dtype: float64

In [11]:
flux = df.HeatFlux

In [13]:
flux.head()

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [14]:
flux[:5]

0    271.8
1    264.0
2    238.8
3    230.7
4    251.6
Name: HeatFlux, dtype: float64

In [18]:
factor1 = df.East
factor3 = df.North
factor2 = df.South

In [19]:
factor_table = np.column_stack((factor1,factor2,factor3))

In [20]:
factor_table

array([[33.53, 40.55, 16.66],
       [36.5 , 36.19, 16.46],
       [34.66, 37.31, 17.66],
       [33.13, 32.52, 17.5 ],
       [35.75, 33.71, 16.4 ],
       [34.46, 34.14, 16.28],
       [34.6 , 34.85, 16.06],
       [35.38, 35.89, 15.93],
       [35.85, 33.53, 16.6 ],
       [35.68, 33.79, 16.41],
       [35.35, 34.72, 16.17],
       [35.04, 35.22, 15.92],
       [34.07, 36.5 , 16.04],
       [32.2 , 37.6 , 16.19],
       [34.32, 37.89, 16.62],
       [31.08, 37.71, 17.37],
       [35.73, 37.  , 18.12],
       [34.11, 36.76, 18.53],
       [34.79, 34.62, 15.54],
       [35.77, 35.4 , 15.7 ],
       [36.44, 35.96, 16.45],
       [37.82, 36.26, 17.62],
       [35.07, 36.34, 18.12],
       [35.26, 35.9 , 19.05],
       [35.56, 31.84, 16.51],
       [35.73, 33.16, 16.02],
       [36.46, 33.83, 15.89],
       [36.26, 34.89, 15.83],
       [37.2 , 36.27, 16.71]])

In [22]:
x = sm.add_constant(factor_table, prepend=True)

In [23]:
x

array([[ 1.  , 33.53, 40.55, 16.66],
       [ 1.  , 36.5 , 36.19, 16.46],
       [ 1.  , 34.66, 37.31, 17.66],
       [ 1.  , 33.13, 32.52, 17.5 ],
       [ 1.  , 35.75, 33.71, 16.4 ],
       [ 1.  , 34.46, 34.14, 16.28],
       [ 1.  , 34.6 , 34.85, 16.06],
       [ 1.  , 35.38, 35.89, 15.93],
       [ 1.  , 35.85, 33.53, 16.6 ],
       [ 1.  , 35.68, 33.79, 16.41],
       [ 1.  , 35.35, 34.72, 16.17],
       [ 1.  , 35.04, 35.22, 15.92],
       [ 1.  , 34.07, 36.5 , 16.04],
       [ 1.  , 32.2 , 37.6 , 16.19],
       [ 1.  , 34.32, 37.89, 16.62],
       [ 1.  , 31.08, 37.71, 17.37],
       [ 1.  , 35.73, 37.  , 18.12],
       [ 1.  , 34.11, 36.76, 18.53],
       [ 1.  , 34.79, 34.62, 15.54],
       [ 1.  , 35.77, 35.4 , 15.7 ],
       [ 1.  , 36.44, 35.96, 16.45],
       [ 1.  , 37.82, 36.26, 17.62],
       [ 1.  , 35.07, 36.34, 18.12],
       [ 1.  , 35.26, 35.9 , 19.05],
       [ 1.  , 35.56, 31.84, 16.51],
       [ 1.  , 35.73, 33.16, 16.02],
       [ 1.  , 36.46, 33.83, 15.89],
 

In [26]:
results = sm.OLS(flux,x).fit()

In [27]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               HeatFlux   R-squared:                       0.874
Model:                            OLS   Adj. R-squared:                  0.859
Method:                 Least Squares   F-statistic:                     57.87
Date:                Fri, 02 Jul 2021   Prob (F-statistic):           2.17e-11
Time:                        12:39:42   Log-Likelihood:                -101.39
No. Observations:                  29   AIC:                             210.8
Df Residuals:                      25   BIC:                             216.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        389.1659     66.094      5.888      0.000     253.043     525.288
x1             2.1247      1.214      1.750      0.092      -0.376       4.626
x2             5.3185      0.963      5.523      0.000       3.335       7.302
x3           -24.1324      1.869    -12.915      0.000     -27.981     -20.284
==============================================================================
Omnibus:                        1.833   Durbin-Watson:                   1.484
Prob(Omnibus):                  0.400   Jarque-Bera (JB):                0.778
Skew:                           0.328   Prob(JB):                        0.678
Kurtosis:                       3.462   Cond. No.                     2.18e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.18e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

heatflux = 2.1 * East + 5.3 * North - 24.1 * South + 389.16

In [1]:
salary = [34.7,28.9,26.7,32.8,43.6,37.9,28.7]
marks = [67.8,59.3,54.7,65.8,78.6,72.9,58.9]

In [6]:
stats.linregress(marks,salary)

LinregressResult(slope=0.699363979179945, intercept=-12.429814637773546, rvalue=0.9928474346463362, pvalue=8.278321576159311e-06, stderr=0.037610030299421816, intercept_stderr=2.4783774749114253)

y = 0.699 * x - 12.4298

IF THE REGRESSION LINE OFTWO VARIABLES MAKES AN ACUTE ANGLE WITH THE X-AXIS IT INDICATES A POSITIVE CORRELATION AND THE SLOP OFTHE LINE WILL BE POSITIVE

R squared is the correlation coefficient.
The sign of the correlation coefficient shows whether the correlation is positive or negative, a positive sign shows that there's a positive correlation and a negative sign shows that the correlation is negative.
- Positive correlation means the variables are directly proportional
- Negative correlation means the variables are inversely proportional

The closer the mod of correlation is to 1 the stronger the correlation is

i.e as |correlation coefficient| approaches 1 the correlation becomes stronger

standard error = standard deviation / sqrt(n)

it indicates the uncertainity in the prediction of the mean on the basis of the sample data

Ho response does not depend on predictor

In [4]:
df.head()

HeatFlux   East  South  North
0     271.8  33.53  40.55  16.66
1     264.0  36.50  36.19  16.46
2     238.8  34.66  37.31  17.66
3     230.7  33.13  32.52  17.50
4     251.6  35.75  33.71  16.40

In [5]:
resp = list(df.HeatFlux)

In [6]:
east = list(df.East)
south = list(df.South)
north = list(df.North)

In [11]:
x = np.column_stack((east,north,south))
x = sm.add_constant(x, prepend = True)

In [12]:
results = sm.OLS(resp,x).fit()

In [13]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.874
Model:                            OLS   Adj. R-squared:                  0.859
Method:                 Least Squares   F-statistic:                     57.87
Date:                Tue, 06 Jul 2021   Prob (F-statistic):           2.17e-11
Time:                        12:51:15   Log-Likelihood:                -101.39
No. Observations:                  29   AIC:                             210.8
Df Residuals:                      25   BIC:                             216.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        389.1659     66.094      5.888      0.000     253.043     525.288
x1             2.1247      1.214      1.750      0.092      -0.376       4.626
x2           -24.1324      1.869    -12.915      0.000     -27.981     -20.284
x3             5.3185      0.963      5.523      0.000       3.335       7.302
==============================================================================
Omnibus:                        1.833   Durbin-Watson:                   1.484
Prob(Omnibus):                  0.400   Jarque-Bera (JB):                0.778
Skew:                           0.328   Prob(JB):                        0.678
Kurtosis:                       3.462   Cond. No.                     2.18e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.18e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

There is a 95% chance that the const value will fall between 253.043 and 525.288

As the interval reduces r-square increases, i.e model becomes more accurate

### Equation
y = m1x1 + m2x2 + m3x3 + c

Heatflux = 2.12 * east - 24.13 * north + 5.31 * south + 389.16

## Other method

In [3]:
import sklearn

In [33]:
from sklearn import linear_model
from sklearn.model_selection import train_test_split

In [5]:
df.columns

Index(['HeatFlux', 'East', 'South', 'North'], dtype='object')

In [44]:
x_train, x_test, y_train, y_test = train_test_split(df[['East','South','North']],df['HeatFlux'],test_size = 0.30, random_state = 45)

In [45]:
x_train.head()

East  South  North
23  35.26  35.90  19.05
6   34.60  34.85  16.06
25  35.73  33.16  16.02
18  34.79  34.62  15.54
17  34.11  36.76  18.53

In [46]:
model = linear_model.LinearRegression()

In [47]:
model.fit(x_train,y_train)

LinearRegression()

In [48]:
model.predict(x_test)

array([264.71861099, 271.22059752, 238.56181156, 257.3579877 ,
       252.75703802, 250.28913652, 219.03096573, 259.85262505,
       266.67448733])

In [49]:
y_test

26    263.0
7     266.5
24    227.5
28    263.8
5     257.9
9     239.3
22    224.7
10    258.0
13    267.0
Name: HeatFlux, dtype: float64

###### Residual = Predicted value - Actual value

In [50]:
resd = list(np.subtract(model.predict(x_test), y_test))
resd_sq = [i**2 for i in resd]

In [51]:
resd

[1.7186109895577033,
 4.72059752017185,
 11.061811563817344,
 -6.44201230147462,
 -5.142961975594062,
 10.989136515776124,
 -5.669034272722001,
 1.852625050879908,
 -0.32551267343058043]

In [52]:
resd_sq

[2.953623733428508,
 22.28404094745262,
 122.3636750734031,
 41.49952249235032,
 26.450057882406377,
 120.76112136236422,
 32.137949585296674,
 3.4322195791477816,
 0.10595850056392371]

In [53]:
sum(resd_sq)

371.98816915641356

In [54]:
np.mean(resd)

1.4181400463312963

## Exercise

In [55]:
salary = [34.7,28.9,26.7,32.8,43.6,37.9,28.7]
marks = [67.8,59.3,54.7,65.8,78.6,72.9,58.9]

In [63]:
X_train, X_test, Y_train, Y_test = train_test_split(np.array(marks),np.array(salary),test_size = 0.30)

In [65]:
model.fit(X_train.reshape(-1,1),Y_train.reshape(-1,1))

LinearRegression()

In [68]:
model.predict(X_test.reshape(-1,1))

array([[28.25644352],
       [33.40323791],
       [25.12361215]])

In [70]:
resd = list(np.subtract(model.predict(X_test.reshape(-1,1)), Y_test.reshape(-1,1)))
resd_sq = [i**2 for i in resd]

In [71]:
resd

[array([-0.44355648]), array([0.60323791]), array([-1.57638785])]

In [72]:
resd_sq

[array([0.19674235]), array([0.36389598]), array([2.48499866])]

In [75]:
float(sum(resd_sq))

3.0456369953098923

# Multi-Collinearity

# 15/7

In [5]:
import pandas as pd
import sklearn
from sklearn.preprocessing import StandardScaler

In [3]:
df = pd.read_csv('Salary.csv')

In [4]:
df.head()

Exp  Salary
0  1.1   39343
1  1.3   46205
2  1.5   37731
3  2.0   43525
4  2.2   39891

In [6]:
scaler = StandardScaler()

In [9]:
scaled_df = pd.DataFrame(scaler.fit_transform(df))

In [11]:
scaled_df.head()

0         1
0 -1.460399 -1.407028
1 -1.404322 -1.190560
2 -1.348245 -1.457879
3 -1.208053 -1.275103
4 -1.151977 -1.389740

In [12]:
np.mean(scaled_df[0])

-3.1720657846433045e-16

In [13]:
np.std(scaled_df[1])

1.0

In [14]:
np.mean(scaled_df[1])

-1.8397981550931167e-16

In [15]:
np.std(scaled_df[1],ddof = 1)

1.0145993123917847

In [16]:
from sklearn.preprocessing import MinMaxScaler

In [17]:
scaler2 = MinMaxScaler(feature_range = (-1,1))

In [20]:
scaled_minmax = pd.DataFrame(scaler2.fit_transform(df))

In [21]:
scaled_minmax.head()

0         1
0 -1.000000 -0.968310
1 -0.967742 -0.833409
2 -0.935484 -1.000000
3 -0.854839 -0.886095
4 -0.822581 -0.957536

## Bank data

In [22]:
df = pd.read_excel('BankLoan.xlsx')

In [23]:
df.head()

Married  Dependents     Education Self_Employed  ApplicantIncome  \
0     Yes           1      Graduate            No             4583   
1     Yes           0      Graduate           Yes             3000   
2     Yes           0  Not Graduate            No             2583   
3      No           0      Graduate            No             6000   
4     Yes           2      Graduate           Yes             5417   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0             1508.0         128               360               1   
1                0.0          66               360               1   
2             2358.0         120               360               1   
3                0.0         141               360               1   
4             4196.0         267               360               1   

  Property_Area  Loan_Amt  
0         Rural         0  
1         Urban         8  
2         Urban        12  
3         Urban        15  
4         Urban        10

In [25]:
df.shape

(492, 11)

In [26]:
df.columns

Index(['Married', 'Dependents', 'Education', 'Self_Employed',
       'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Amt'],
      dtype='object')

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 492 entries, 0 to 491
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Married            492 non-null    object 
 1   Dependents         492 non-null    int64  
 2   Education          492 non-null    object 
 3   Self_Employed      492 non-null    object 
 4   ApplicantIncome    492 non-null    int64  
 5   CoapplicantIncome  492 non-null    float64
 6   LoanAmount         492 non-null    int64  
 7   Loan_Amount_Term   492 non-null    int64  
 8   Credit_History     492 non-null    int64  
 9   Property_Area      492 non-null    object 
 10  Loan_Amt           492 non-null    int64  
dtypes: float64(1), int64(6), object(4)
memory usage: 42.4+ KB


In [29]:
x = df.drop('Loan_Amt',axis = 1)

In [30]:
x.head()

Married  Dependents     Education Self_Employed  ApplicantIncome  \
0     Yes           1      Graduate            No             4583   
1     Yes           0      Graduate           Yes             3000   
2     Yes           0  Not Graduate            No             2583   
3      No           0      Graduate            No             6000   
4     Yes           2      Graduate           Yes             5417   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0             1508.0         128               360               1   
1                0.0          66               360               1   
2             2358.0         120               360               1   
3                0.0         141               360               1   
4             4196.0         267               360               1   

  Property_Area  
0         Rural  
1         Urban  
2         Urban  
3         Urban  
4         Urban

In [31]:
y = df.Loan_Amt

In [37]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.25)

In [43]:
x_train.head()

Married  Dependents Education Self_Employed  ApplicantIncome  \
300     Yes           2  Graduate            No             5391   
100     Yes           0  Graduate           Yes             3459   
169     Yes           3  Graduate            No             3430   
341     Yes           0  Graduate            No             2920   
388     Yes           1  Graduate            No            18333   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
300           0.000000         130               360               1   
100           0.000000          25               120               1   
169        1250.000000         128               360               0   
341          16.120001          87               360               1   
388           0.000000         500               360               1   

    Property_Area  
300         Urban  
100     Semiurban  
169     Semiurban  
341         Rural  
388         Urban

In [39]:
x_train.shape

(369, 10)

In [40]:
x_train_numcols = x_train.select_dtypes(exclude=object)

In [41]:
x_train_numcols.head()

Dependents  ApplicantIncome  CoapplicantIncome  LoanAmount  \
300           2             5391           0.000000         130   
100           0             3459           0.000000          25   
169           3             3430        1250.000000         128   
341           0             2920          16.120001          87   
388           1            18333           0.000000         500   

     Loan_Amount_Term  Credit_History  
300               360               1  
100               120               1  
169               360               0  
341               360               1  
388               360               1

In [42]:
x_train_numcols.shape

(369, 6)

In [44]:
x_train_chrcol = x_train.select_dtypes(include = object)

In [45]:
x_train_chrcol.head()

Married Education Self_Employed Property_Area
300     Yes  Graduate            No         Urban
100     Yes  Graduate           Yes     Semiurban
169     Yes  Graduate            No     Semiurban
341     Yes  Graduate            No         Rural
388     Yes  Graduate            No         Urban

In [46]:
x_train_chrcol.shape

(369, 4)